cluster-autoscaler/core/scale_down.go

```diff
@@ -146,6 +177,49 @@ func calculateCoresAndMemoryTotal(nodes []*apiv1.Node, timestamp time.Time) (int
 	return coresTotal, memoryTotal
 }
 
+func calculateGpusTotal(nodes []*apiv1.Node, cp cloudprovider.CloudProvider, timestamp time.Time) (map[string]int64, error) {
+	type gpuInfo struct {
+		name  string
+		count int64
+	}
+
+	result := make(map[string]int64)
+	ngCache := make(map[string]gpuInfo)
+	for _, node := range nodes {
+		if isNodeBeingDeleted(node, timestamp) {
+			// Nodes being deleted do not count towards total cluster resources
+			continue
+		}
+		nodeGroup, err := cp.NodeGroupForNode(node)
+		if err != nil {
+			return nil, fmt.Errorf("can not get node group for node %v when calculating cluster gpu usage (%s)", node.Name, err.Error())
+		}
+		if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
+			return nil, fmt.Errorf("group not found for node %v when calculating cluster gpu usage", node.Name)
```

Comment:
I think we should log and continue here. NodeGroupForNode() returns nil for every non-autoscaled node (explicitly said here: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L44). Having non-autoscaled nodes is not an error condition and shouldn't prevent autoscaling from working.

Reply:
I thought about it some more. The problem is as follows: NodeGroup is needed to predict the number of GPUs on a node that is currently installing GPU driver. We don't have node NodeGroup for non autoscaled node and so we won't be able to guess how many GPUs it will have once drivers are installed. We can't fix this without some large refactors, that I disagree with both in principle and for practical reasons. We can still get the number from Capacity once the driver is installed.

The simplest thing to do is to change GetNodeTargetGpus to accept a nil NodeGroup. It would than return an error if it actually got to the point where it needs to call TemplateNodeInfo and NodeGroup is nil. In most cases we would be perfectly fine. During driver installation on non-autoscaled Node we would return an error resulting in setting limitUnknown and effectively stopping GPU autoscaling until the driver is installed. It's not perfect, but the scenario is unlikely and I agreed with @mwielgus that it's acceptable as a temporary limitation.

Also it's gpu.go, which is already the hack el dorado.

Anyway - WDYT? Do you see anything I missed?

Finally if you do end up implementing that idea, please return an InternalError if the hack is triggered (AutoscalerError is still an error and ToAutoscalerError does type assertion and only changes type if it's not an AutoscalerError).

Reply:
seems reasonable. Unless sth is screwed up which would imply that node will never get the gpu capacity (even though it contain gpu label) the situation should be self healing.